java/gradle.htmlui/src/org/netbeans/modules/gradle/htmlui/HtmlJavaPagesNodeFactory.java

```diff
@@ -0,0 +1,110 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.netbeans.modules.gradle.htmlui;
+
+import java.awt.Image;
+import org.netbeans.modules.gradle.api.NbGradleProject;
+import org.netbeans.modules.gradle.spi.nodes.AbstractGradleNodeList;
+import java.beans.PropertyChangeEvent;
+import java.beans.PropertyChangeListener;
+import java.util.Collections;
+import java.util.List;
+import javax.swing.event.ChangeListener;
+import org.netbeans.api.project.Project;
+import org.netbeans.spi.project.ui.support.NodeFactory;
+import org.netbeans.spi.project.ui.support.NodeList;
+import org.openide.filesystems.FileObject;
+import org.openide.loaders.DataFolder;
+import org.openide.nodes.FilterNode;
+import org.openide.nodes.Node;
+import org.openide.util.ImageUtilities;
+import org.openide.util.NbBundle;
+
+@NbBundle.Messages({
+    "MSG_HtmlJavaPages=Frontend UI Pages"
+})
+@NodeFactory.Registration(projectType = NbGradleProject.GRADLE_PROJECT_TYPE, position = 77)
+public final class HtmlJavaPagesNodeFactory implements NodeFactory {
+
+    @Override
+    public NodeList<?> createNodes(Project p) {
+        return new HtmlJavaPagesList(p);
+    }
+
+    private static class HtmlJavaPagesList extends AbstractGradleNodeList<FileObject> implements PropertyChangeListener {
+
+        final Project project;
+
+        HtmlJavaPagesList(Project project) {
+            this.project = project;
+        }
+
+        @Override
+        public List<FileObject> keys() {
+            FileObject pages = project.getProjectDirectory().getFileObject("src/main/webapp/pages"); // NOI18N
+            return pages != null ? Collections.singletonList(pages) : Collections.<FileObject>emptyList();
+        }
+
+        @Override
+        public Node node(FileObject key) {
+            DataFolder df = DataFolder.findFolder(key);
+            FilterNode fn = new FilterNode(df.getNodeDelegate().cloneNode()) {
+                @Override
+                public Image getIcon(int type) {
+                    return ImageUtilities.loadImage("org/netbeans/modules/gradle/htmlui/DukeHTML.png"); // NOI18N
+                }
+
+                @Override
+                public Image getOpenedIcon(int type) {
+                    return getIcon(type);
+                }
+
+                @Override
+                public String getName() {
+                    return "pages"; // NOI18N
+                }
+
+                @Override
+                public String getDisplayName() {
+                    return Bundle.MSG_HtmlJavaPages();
+                }
+            };
+            return fn;
+        }
+
+        @Override
+        public void propertyChange(PropertyChangeEvent evt) {
+            if (NbGradleProject.PROP_PROJECT_INFO.equals(evt.getPropertyName())) {
```

Comment:
This would invoke change if the Gradle project files would be changed. Maybe the intent was to refresh on PROP_RESOURCES changes. Though in order that one to work you need to register a WatchedResourceProvider for "src/main/webapp/pages". In that case evt.getNewValue() will hold the URI of the changed resource.

Reply:
I admit I had no intent, I just copied [WebPagesNodeFactory](https://github.com/apache/netbeans/blob/master/enterprise/gradle.javaee/src/org/netbeans/modules/gradle/javaee/nodes/WebPagesNodeFactory.java#L72). Dump, complete (e.g. not even **selective**) [cluelessness](http://wiki.apidesign.org/wiki/Cluelessness)...

Reply:
I guess  you are not the only one [doing that](https://github.com/apache/netbeans/blob/0719659fdb6d42a5d2314dfd0436a50c6ef81c77/enterprise/maven.j2ee/src/org/netbeans/modules/maven/j2ee/ui/nodes/WebPagesNodeFactory.java#L108)...